### PR TITLE
[GraphBolt] Implement `InSubgraphSampler`

### PR DESCRIPTION
**Pull request by Skeleton003**
## Description

Issue #6306 
Added implementation of `InSubgraphSampler`, a subclass of `SubgraphSampler`, as wll as the test code.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/impl/in_subgraph_sampler.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+from ..subgraph_sampler import SubgraphSampler
+from ..utils import unique_and_compact_node_pairs
+from .sampled_subgraph_impl import FusedSampledSubgraphImpl
+
+
+__all__ = ["InSubgraphSampler"]
+
+
+@functional_datapipe("sample_in_subgraph")
+class InSubgraphSampler(SubgraphSampler):
```
Comment (keli-wen) [inline python/dgl/graphbolt/impl/in_subgraph_sampler.py:14]: Should we add the Fused prefix, considering the future introduction of `Non-Fused SampledSubgraph`?

Comment (Skeleton003) [reply python/dgl/graphbolt/impl/in_subgraph_sampler.py:14]: I'm not familiar with that. But why not do the renaming job after the future introduction of `Non-Fused SampledSubgraph`?

Comment (frozenbugs) [reply python/dgl/graphbolt/impl/in_subgraph_sampler.py:14]: It is not necessary to rename it, the InSubgraphSampler should not care about which SamplingGraph is used.

File: tests/python/pytorch/graphbolt/impl/test_in_subgraph_sampler.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import dgl.graphbolt as gb
+import gb_test_utils
+import pytest
+import torch
+
+
+def test_InSubgraphSampler_homo():
+    num_seeds = 100
+    item_set = gb.ItemSet(torch.arange(0, num_seeds), names="seed_nodes")
+    graph = gb_test_utils.rand_csc_graph(num_seeds, density=0.1)
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_in_subgraph_sampler.py:10]: The graph should have significantly more nodes than just num_seed nodes.

File: tests/python/pytorch/graphbolt/impl/test_in_subgraph_sampler.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import dgl.graphbolt as gb
+import gb_test_utils
+import pytest
+import torch
+
+
+def test_InSubgraphSampler_homo():
+    num_seeds = 100
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_in_subgraph_sampler.py:8]: Let's make it smaller, say 5 should be enough.

File: tests/python/pytorch/graphbolt/impl/test_in_subgraph_sampler.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import dgl.graphbolt as gb
+import gb_test_utils
+import pytest
+import torch
+
+
+def test_InSubgraphSampler_homo():
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_in_subgraph_sampler.py:7]: better to add test case for hetero case.

File: python/dgl/graphbolt/impl/in_subgraph_sampler.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    """Sample neighbor edges from a graph and return a subgraph.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/in_subgraph_sampler.py:15]:
```suggestion
    """Sample the subgraph induced on the inbound edges of the given nodes.
```

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**